File: algorithms/python/验证二叉搜索树/验证二叉搜索树.py
# Source : https://leetcode.cn/problems/validate-binary-search-tree/description/
# Author : Yanan Wang
# Date   : 2023-08-30

# 二叉搜索树的中序遍历是严格递增的，按照这一规律，可以设置一个全局变量prev
# 当在遍历当前节点的时候，prev一定是严格小于当前节点的，否则返回False

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        prev = float('-inf')
        def dfs(root):
            if not root:
                return True
            if not dfs(root.left):
                return False
            nonlocal prev
            if prev >= root.val:
                return False
            prev = root.val
            if not dfs(root.right):
                return False
            return True

        return dfs(root)

File: algorithms/cpp/验证二叉搜索树/验证二叉搜索树.cpp
// Source : https://leetcode.cn/problems/validate-binary-search-tree/description/ 
// Author : Yanan Wang
// Date   : 2023-11-28

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long prev = (long long)INT_MIN - 1;
    bool isValidBST(TreeNode* root) {
        if (root == nullptr) {
            return true;
        }
        if (!isValidBST(root->left)) {
            return false;
        }
        if (root->val <= prev) {
            return false;
        } 
        prev = root->val;
        if (!isValidBST(root->right)) {
            return false;
        } 
        return true;
    }
};